model/map/src/main/java/org/keycloak/models/map/events/MapEventStoreProvider.java

```diff
@@ -0,0 +1,181 @@
+import org.keycloak.models.map.storage.criteria.DefaultModelCriteria;
+
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import static org.keycloak.common.util.StackUtil.getShortStackTrace;
+import static org.keycloak.models.map.events.EventUtils.modelToEntity;
+
+public class MapEventStoreProvider implements EventStoreProvider {
+
+    private static final Logger LOG = Logger.getLogger(MapEventStoreProvider.class);
+    private final KeycloakSession session;
+    private final MapKeycloakTransaction<MapAuthEventEntity, Event> authEventsTX;
+    private final MapKeycloakTransaction<MapAdminEventEntity, AdminEvent> adminEventsTX;
+
+    public MapEventStoreProvider(KeycloakSession session, MapStorage<MapAuthEventEntity, Event> loginEventsStore, MapStorage<MapAdminEventEntity, AdminEvent> adminEventsStore) {
+        this.session = session;
+        this.authEventsTX = loginEventsStore.createTransaction(session);
+        this.adminEventsTX = adminEventsStore.createTransaction(session);
+
+        session.getTransactionManager().enlistAfterCompletion(this.authEventsTX);
+        session.getTransactionManager().enlistAfterCompletion(this.adminEventsTX);
+    }
+
+    /** LOGIN EVENTS **/
+    @Override
+    public void onEvent(Event event) {
+        LOG.tracef("onEvent(%s)%s", event, getShortStackTrace());
+        String realmId = event.getRealmId();
+        String id = event.getId();
+        if (realmId == null) {
+            throw new IllegalStateException("RealmId was null for Event " + id);
+        }
+
+        RealmModel realm = session.realms().getRealm(realmId);
+        if (realm == null) {
+            throw new IllegalStateException("Realm with id " + realmId + " not found.");
+        }
+
+        if (id != null && authEventsTX.read(id) != null) {
+            throw new ModelDuplicateException("Event already exists: " + id);
+        }
+
+        MapAuthEventEntity e = modelToEntity(event);
+        e.setExpiration(TimeAdapter.fromIntegerWithTimeInSecondsToLongWithTimeAsInSeconds(Time.currentTime()) + realm.getEventsExpiration());
+        authEventsTX.create(e);
+    }
+
+    private boolean filterExpired(ExpirableEntity event) {
+        Long expiration = event.getExpiration();
+        // Check if entity is expired
+        if (expiration == null || expiration < TimeAdapter.fromIntegerWithTimeInSecondsToLongWithTimeAsInSeconds(Time.currentTime())) {
+            // Remove entity
+            authEventsTX.delete(event.getId());
+
+            return false; // Do not include entity in the resulting stream
+        }
+
+        return true; // Entity is not expired
+    }
+
+    @Override
+    public EventQuery createQuery() {
+        LOG.tracef("createQuery()%s", getShortStackTrace());
+        return new MapAuthEventQuery(((Function<QueryParameters<Event>, Stream<MapAuthEventEntity>>) authEventsTX::read)
+                .andThen(s -> s.filter(this::filterExpired).map(EventUtils::entityToModel)));
```

Comment:
Does this obtain all events from the DB and only after obtaining them, filters them via `filterExpired` method?
If that is so, it needs to be changed so that the query always incorporates a searchable field for expiration so that the expired events do not get into the result set

Reply:
No, this should return only not-expired events. I added this check just to double-check expiration also in Java. It is added here:
https://github.com/keycloak/keycloak/pull/11822/files#diff-fc8f503c6f676186c84a79c94d577e37f2db8ea747919f1ee813418a93a911adR110